src/main/java/sorald/event/collectors/MinerStatisticsCollector.java

```diff
@@ -0,0 +1,63 @@
+package sorald.event.collectors;
+
+import java.util.*;
+import sorald.event.SoraldEvent;
+import sorald.event.SoraldEventHandler;
+import sorald.event.models.miner.MinedRuleEvent;
+
+/** Event handler for recording the miner mode statistics. */
+public class MinerStatisticsCollector implements SoraldEventHandler {
+    /** start of mining determined by currentTimeMillis */
+    private long miningStartTime;
+
+    /** start of mining determined by currentTimeMillis */
+    private long miningEndTime;
+
+    private List<MinedRuleEvent> minedRules = new ArrayList<>();
+
+    @Override
+    public void registerEvent(SoraldEvent event) {
+        switch (event.type()) {
+            case MINING_START:
+                miningStartTime = System.currentTimeMillis();
+                break;
+            case MINING_END:
+                miningEndTime = System.currentTimeMillis();
+                break;
+            case MINED:
+                MinedRuleEvent minedRuleEvent = (MinedRuleEvent) event;
+                Optional<MinedRuleEvent> existingEvent =
+                        minedRules.stream()
+                                .filter(x -> x.getRuleKey().equals(minedRuleEvent.getRuleKey()))
+                                .findFirst();
+
+                if (!existingEvent.isEmpty()) {
+                    existingEvent.get().addWarningLocations(minedRuleEvent.getWarningLocations());
+                } else {
+                    minedRules.add(minedRuleEvent);
+                }
```

Comment:
This I would advice against. Events should be immutable. If another event handler has also gotten this particular event, its caught event would be mutate as well, which would be a very high-risk side effect.

My advice: make `MinedRuleEvent` immutable and have it only contain which event was mined. Then keep track of the amounts of events in an internal data structure in this event handler.

Reply:
Done! (have a look to see if it needs further fixes)

Reply:
Actually, this one is also related to the event-handler architecture. Maybe we can make it simpler by changing the architecture.

Reply:
I don't think it is related to the architecture. Only the casting is. I mean, we could send a _copy_ of each event to each handler, but that would entail having to make a copy constructor for each event as well, so it would be much easier to just keep to immutable events.

I mean, what would be architectural is to actually enforce immutable events. I don't think there's any way to enforce immutability natively in Java. We could make a Spoon processor to check it. Any ideas?